contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,278 @@
+            // Calculate the maximum possible reward per service during the last deposit period
+            uint256 maxRewardsPerService = (rewardsPerSecond * (block.timestamp - tsLast)) / numServices;
+            // Traverse all the eligible services and calculate their rewards
+            for (uint256 i = 0; i < numServices; ++i) {
+                uint256 curServiceId = eligibleServiceIds[i];
+                ServiceInfo storage curInfo = mapServiceInfo[curServiceId];
+
+                // Calculate the reward up until now
+                // If the staking was longer than the deposited period, the service's timestamp is adjusted such that
+                // it is equal to at most the tsLastBalance of the last deposit happening during every _checkpoint() call
+                uint256 reward = rewardsPerSecond * (block.timestamp - curInfo.ts);
+                // Adjust the reward if it goes out of calculated max bounds
+                if (reward > maxRewardsPerService) {
+                    reward = maxRewardsPerService;
+                }
+
+                // Adjust the deposit balance
+                if (lastBalance >= reward) {
+                    lastBalance -= reward;
+                } else {
+                    // This situation must never happen
+                    // TODO: Fuzz this
+                    reward = lastBalance;
+                    lastBalance = 0;
+                }
+
+                // Add the calculated reward to the service info
+                curInfo.reward += reward;
+
+                // Adjust the starting ts for each service to a current timestamp
+                curInfo.ts = block.timestamp;
+            }
+            // Update the storage balance
+            balance = lastBalance;
+        }
+
+        // Record the current timestamp such that next calculations start from this point of time
+        tsLastBalance = block.timestamp;
+
+        emit Checkpoint(lastBalance);
+    }
+
+    /// @dev Unstakes the service.
+    /// @param serviceId Service Id.
+    function unstake(uint256 serviceId) external {
+        ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+        // Check for the service ownership
+        if (msg.sender != sInfo.owner) {
+            revert();
+        }
+
+        // Call the checkpoint and get the service index in the set of services
+        uint256 idx = _checkpoint(serviceId);
+
+        // Transfer the service back to the owner
+        IService(serviceRegistry).transferFrom(address(this), msg.sender, serviceId);
+
+        // Send the remaining small balance along with the reward if it is below the chosen threshold
+        uint256 amount = sInfo.reward;
+        uint256 lastBalance = balance;
+        if (lastBalance < minBalance) {
+            amount += lastBalance;
+            balance = 0;
+        }
+
+        // Transfer accumulated rewards to the service owner
+        if (amount > 0) {
+            _withdraw(msg.sender, amount);
```

Comment:
Needs to be sent to multisig, not service owner!

Reply:
I understand correctly that receiving a reward is not a mirror (asymmetrical) operation to "staking"

Reply:
Yes the service logic is responsible for deciding where rewards go.